### PR TITLE
Update jupyter notebook and create training script + streamlit

### LINKED ISSUE #7: Train 1st model using Weak Labels
Effort: 1 person (max 2)

We created a labeled dataset created with Weak Supervision available here: https://huggingface.co/datasets/bergr7/weakly_supervised_ag_news

We should now train the first model on that and save it to a place accessible for the rest of the team.

Suggested steps:
- fine-tune `distilbert`
- train using huggingface Transformers API
- track experiment in W&B: [docs](https://docs.wandb.ai/guides/integrations/huggingface)
- store the Model in the W&B model registry and/or hugging face hub

We should have a team space in W&B if I am not wrong.

### LINKED ISSUE #10: Create model UI for users
Effort: 1 person (max 2)

We should create a simple UI for users to query and interact with the model.
The UI should require an input text from the user, request the model prediction using the model API endpoint in [`services/api-serveless`](https://github.com/EdAbati/fsdl-2022-weak-supervision-project/tree/main/services/api-serverless) and show the prediction in an informative way.

Ideally it should be setup to run in a docker container (or AWS EC2 like the example course labs?).

Options for implementation:
- Gradio
- Streamlit

### PR DESCRIPTION
**Pull request by diegoquintanav**
this contains a basic training script that uses huggingface + wandb. To make it work, update the .env file with `HF_TOKEN` and `WANDB_API_TOKEN`. I've provided my tokens in the lambdalabs machine in `/playground/repo/fsdl-2022-yadayada/.env`.

This should close #7, but it needs another pair of eyes. I only refactored a script @kushal had in his directory.

Among other things, this PR

1. changes usage of anaconda for poetry
2. creates a package called `app` in the service, for packaging scripts

the image for this container is very heavy :|

Closes #10

### CODE REVIEW

**Review by diegoquintanav — COMMENTED** [review]
I'd rather remove the commented lines

File: services/api-serverless/api/requirements.txt [hunk 1]
```diff
@@ -1 +0,0 @@
-numpy==1.23.3
```
Comment (EdAbati) [inline services/api-serverless/api/requirements.txt]: Please revert changes this will be done in a separate PR

File: services/lambda/api/Dockerfile [hunk 2]
```diff
@@ -1,6 +1,11 @@
 FROM public.ecr.aws/lambda/python:3.8
```
Comment (EdAbati) [inline services/lambda/api/Dockerfile]: Please revert the changes of the whole `api` folder. This will be done is a separate PR.

File: services/streamlit/README.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# About
```
Comment (EdAbati) [inline services/streamlit/README.md:1]: This README is still a placeholder

File: services/jupyter/src/app/config.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import secrets
+from typing import Optional
+
+from pydantic import AnyHttpUrl, BaseSettings, HttpUrl, validator
+
+
+class SettingsModeEnum(str, enum.Enum):
+    DEV = "dev"
+    PROD = "prod"
+    TEST = "test"
```
Comment (EdAbati) [inline services/jupyter/src/app/config.py:10]:
```suggestion
```
Remove unused

File: services/jupyter/src/app/model/train_model.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        data_files=labeled_data_files,
+        features=labeled_features,
+        split="test",
+    )
+
+    unlabeled_dataset = load_dataset(
+        dataset_name,
+        data_files=unlabeled_data_files,
+        features=unlabeled_features,
+    )
```
Comment (EdAbati) [inline services/jupyter/src/app/model/train_model.py]:
```suggestion
```

Remove unused

File: services/jupyter/src/app/model/train_model.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    s3 = boto3.client(
+        "s3",
+        settings.AWS_REGION,
+        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
+    )
+
+    # upload file from local directory to s3 bucket
+    with open(torchscript_dir, "rb") as f:
+        s3.upload_fileobj(f, bucket_name, s3_filename)
```
Comment (EdAbati) [inline services/jupyter/src/app/model/train_model.py:175]:
```suggestion
```
Not needed

File: services/jupyter/src/app/model/train_model.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    }
+
+    # load tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
+
+    dummy_tokenized_input = tokenizer(
+        dummy_input["text"], truncation=True, return_tensors="pt"
+    )
+
+    return torch.jit.trace(model, tuple(dummy_tokenized_input.values()))
```
Comment (EdAbati) [inline services/jupyter/src/app/model/train_model.py:248]:
```suggestion
```
Not needed

File: services/jupyter/src/app/model/train_model.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    model_dir = load_model_from_wandb(w, return_dir=True)
+
+    out = convert_model_to_torchscript(
+        model_dir=model_dir,
+        model_ckpt=model_ckpt,
+    )
+
+    out.save("model.pt")
+
+    return out
```
Comment (EdAbati) [inline services/jupyter/src/app/model/train_model.py:270]:
```suggestion
```

Not needed

File: services/jupyter/src/app/config.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+class Settings(BaseSettings):
+
+    PROJECT_NAME: str
+    SENTRY_DSN: Optional[HttpUrl] = None
+
+    @validator("SENTRY_DSN", pre=True)
+    def sentry_dsn_can_be_blank(cls, v: str) -> Optional[str]:
+        if bool(v) is False:
+            return None
+        return v
```
Comment (EdAbati) [inline services/jupyter/src/app/config.py]: I don't think this is needed right?
```suggestion
```

File: services/jupyter/src/app/cli.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    color = typer.colors.BLUE
+    bold = False
+
+    for k, v in settings:
+        if isinstance(v, list):
+            typer.secho(f"{k}:", color=color, bold=bold)
+            for e in v:
+                typer.secho(f"\t{e}", color=color, bold=bold)
+        else:
+            typer.secho(f"{k}: {v}", color=color, bold=bold)
```
Comment (EdAbati) [inline services/jupyter/src/app/cli.py:33]: This could become a method of `Settings`.

Also FYI if you install `typer[all]` if comes with `rich`. Just by using `rich.print`, you get some colours and pretty printing.

File: services/jupyter/src/app/model/train_model.py [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    batch_size: int = 64,
+):
+
+    train_dataset, val_dataset, _ = load_data()
+
+    tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
+
+    def tokenize(batch):
+        return tokenizer(batch["text"], padding=True, truncation=True)
+
```
Comment (EdAbati) [inline services/jupyter/src/app/model/train_model.py:111]: Couple of things here.

1. We could remove `padding=True` from the `tokenize` function and instead add a `data_collator = DataCollatorWithPadding(tokenizer=tokenizer)` that we need to pass to the `Trainer` as extra argument. see: https://huggingface.co/course/chapter3/2?fw=pt#dynamic-padding

2. For each `dataset` (`train_` and `val_`) we could also remove `text` that we don't need (`train_dataset.map(tokenize, batched=True, batch_size=None).remove_columns('text')`) and also set the format to `torch` (`dataset.set_format('torch')`)

File: services/streamlit/src/app/cli.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
+from typing import Optional
```
Comment (EdAbati) [inline services/streamlit/src/app/cli.py:1]: This file is unused can be removed.

File: services/streamlit/src/app/config.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    ALGORITHM: str = "HS256"
+
+    PROJECT_NAME: str
+    SENTRY_DSN: Optional[HttpUrl] = None
+
+    @validator("SENTRY_DSN", pre=True)
+    def sentry_dsn_can_be_blank(cls, v: str) -> Optional[str]:
+        if bool(v) is False:
+            return None
+        return v
```
Comment (EdAbati) [inline services/streamlit/src/app/config.py:28]: is this needed?

File: services/streamlit/src/app/streamlit_app.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+def classify_news(classifier, text):
+    preds = classifier(text, return_all_scores=False)
+    label = {0: "World", 1: "Sports", 2: "Business", 3: "Sci/Tech"}
+    text_label = label.get(int(preds[0]["label"][-1]))
+    preds = classifier(text, return_all_scores=True)
+    for items in preds[0]:
+        items["label"] = label.get(int(items["label"][-1]))
+
+    return text_label, preds[0]
```
Comment (EdAbati) [inline services/streamlit/src/app/streamlit_app.py]: I think these are not needed anymore right?

**Review by EdAbati — CHANGES_REQUESTED** [review]
Hey @diegoquintanav and @ramaiyaKushal , thank you very much :)

I have reviewed, there are mainly things that are not needed anymore and should be deleted.

I have pushed a couple of commits just for cosmetic stuff (removed comments and unused imports)

File: services/api-serverless/api/requirements.txt (hunk 1 above)
Comment (diegoquintanav) [reply services/api-serverless/api/requirements.txt:1]: ok!

File: services/streamlit/README.md (hunk 3 above)
Comment (diegoquintanav) [reply services/streamlit/README.md:1]: What do you mean?

File: services/jupyter/src/app/config.py (hunk 9 above)
Comment (diegoquintanav) [reply services/jupyter/src/app/config.py]: nope! leftovers from other services

Comment (diegoquintanav) [reply services/jupyter/src/app/config.py]: although the settings import is needed for parsing environment variables

File: services/streamlit/src/app/config.py (hunk 13 above)
Comment (diegoquintanav) [reply services/streamlit/src/app/config.py:28]: it's optional, for logging errors in services

File: services/jupyter/src/app/cli.py (hunk 10 above)
Comment (diegoquintanav) [reply services/jupyter/src/app/cli.py:33]: cool, thanks for the tip! I think this belongs to the CLI and not to the settings, since it uses `typer.secho`, so I'm leaving it as it is.

File: services/jupyter/src/app/model/train_model.py (hunk 7 above)
Comment (diegoquintanav) [reply services/jupyter/src/app/model/train_model.py:248]: at some point we will need to migrate the CLI to the jupyter container to not rely on the docker host. I think it can stay since it's not being called.

File: services/jupyter/src/app/model/train_model.py (hunk 6 above)
Comment (diegoquintanav) [reply services/jupyter/src/app/model/train_model.py:175]: at some point we will need to migrate the CLI to the jupyter container to not rely on the docker host. I think it can stay since it's not being called.

File: services/jupyter/src/app/model/train_model.py (hunk 8 above)
Comment (diegoquintanav) [reply services/jupyter/src/app/model/train_model.py:270]: at some point we will need to migrate the CLI to the jupyter container to not rely on the docker host. I think it can stay since it's not being called.

File: services/streamlit/src/app/cli.py (hunk 12 above)
Comment (diegoquintanav) [reply services/streamlit/src/app/cli.py:1]: agree, but it's there in case we need to implement a CLI at the container level

File: services/streamlit/src/app/streamlit_app.py (hunk 14 above)
Comment (diegoquintanav) [reply services/streamlit/src/app/streamlit_app.py]: nope, removing!

File: services/jupyter/src/app/config.py (hunk 4 above)
Comment (diegoquintanav) [reply services/jupyter/src/app/config.py:10]: It can be used in the CLI and at the container level, leaving it as a proposed interface.

File: services/jupyter/src/app/model/train_model.py (hunk 11 above)
Comment (diegoquintanav) [reply services/jupyter/src/app/model/train_model.py:111]: I would do that in another PR, what do you say?

File: services/lambda/api/Dockerfile (hunk 2 above)
Comment (diegoquintanav) [reply services/lambda/api/Dockerfile]: I needed them to test the streamlit interface. Removing them would break the streamlit service, I suggest solving any conflicts when the separate PR is rebased and eventually merged.

File: services/api-serverless/api/requirements.txt (hunk 1 above)
Comment (diegoquintanav) [reply services/api-serverless/api/requirements.txt]: I needed them to test the streamlit interface. Removing them would break the streamlit service, I suggest solving any conflicts when the separate PR is rebased and eventually merged.

File: services/streamlit/README.md (hunk 3 above)
Comment (EdAbati) [reply services/streamlit/README.md:1]: can you add a couple of lines to say what this UI is about? how to use it? screenshot or stuff like that

File: services/jupyter/src/app/model/train_model.py (hunk 11 above)
Comment (EdAbati) [reply services/jupyter/src/app/model/train_model.py:111]: Here is fine I think since it should be about the training script

File: services/jupyter/src/app/model/train_model.py (hunk 7 above)
Comment (EdAbati) [reply services/jupyter/src/app/model/train_model.py:248]: > we will need to migrate the CLI to the jupyter container to not rely on the docker host

Could you elaborate on why you think we should do that? :)

File: services/jupyter/src/app/config.py (hunk 4 above)
Comment (EdAbati) [reply services/jupyter/src/app/config.py:10]: I am tempted to go in the direction of "everything nice to have (but that is not used) should go since we have less than a week left and we need to tidy up the code base"

File: services/streamlit/README.md (hunk 3 above)
Comment (diegoquintanav) [reply services/streamlit/README.md:1]: ah okay

File: services/jupyter/src/app/config.py (hunk 4 above)
Comment (diegoquintanav) [reply services/jupyter/src/app/config.py:10]: It's a nice thing to have indeed. Shall we mark this as resolved then?

File: services/jupyter/src/app/model/train_model.py (hunk 7 above)
Comment (diegoquintanav) [reply services/jupyter/src/app/model/train_model.py:248]: the deploy commands should be accessed from a container. Using the CLI outside of one would force CI to clone the entire repo instead of pulling a docker image. 

I was thinking of using the `jupyter` container for all things model management

**Comment by diegoquintanav** [comment]
I don't understand why the last commit is not passing the checks

File: services/jupyter/src/app/model/train_model.py (hunk 7 above)
Comment (EdAbati) [reply services/jupyter/src/app/model/train_model.py:248]: Mmm I think we can get rid of the Jupiter container instead. but it will be for another task (if we have time)

**Comment by EdAbati** [comment]
Please resolve conflicts @diegoquintanav :)

**Review by EdAbati — APPROVED** [review]
